rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,253 @@
+---
+authors: Andrew Lytvynov (andrew@goteleport.com)
+state: draft
+---
+
+# RFD 15 - 2FA device management
+
+## What
+
+Improvements to 2FA device management (OTP and U2F):
+- support for multiple 2FA devices per account
+- ability to mix OTP and U2F devices (and possibly more in the future)
+- ability to list/add/delete devices by the user from Web UI and CLI
+
+## Why
+
+Currently, Teleport only supports a single OTP or U2F device per local user.
+This setting is global to the cluster - either everyone uses OTP or U2F or
+nothing. Users don't have a choice, and losing a 2FA device requires an admin
+to reset the account to recover.
+
+[RFD 14](0014-session-2FA.md) adds support for 2FA challenges per-connection.
+2FA devices will become more widely used by Enterprise users, who would
+previously delegate this to their SSO provider.
+
+Teleport needs to make 2FA device support more flexible, as both a
+long-standing OSS users' request, and the new Enterprise use-case.
+
+## Details
+
+### protocols
+
+Teleport supports two 2FA protocols: OTP and U2F.
+
+There's no change to the list of supported protocols, but the implementation
+should assume that more protocols will be added (e.g. WebAuthn).
+
+For U2F, we will migrate from https://github.com/tstranex/u2f to
+https://github.com/flynn/u2f, to allow client-side CLI authentication without
+the `u2f-host` dependency.
+
+### UX
+
+For the below examples, we assume that the user already has at least one 2FA
+device enrolled. The [bootstrap](#bootstrap) section covers registration of the
+first device.
+
+#### CLI
+
+Login:
+
+```sh
+$ tsh login --login=non-sso-user
+Enter password for Teleport user non-sso-user: ...
+Tap your security key... <tap>
+
+$ tsh login --login=non-sso-user --mfa=otp
+Enter password for Teleport user non-sso-user: ...
+Enter OTP code: ...
+
+$ tsh login --login=sso-user --auth=oidc
+# SSO page opens
+# no 2FA prompt from Teleport
+```
+
+2FA management:
+
+```sh
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Last used
+----------------   ----   ------------------------------------   -------------------------------
+android OTP        OTP    fa004bf4-acc7-435d-8965-5f5a0a4552e8   Tue 15 Dec 2020 01:29:42 PM PST
```

Comment:
I've done a bit of UX research and most SaaS apps display the **Added Date** instead of the last time of use. Google shows _Added Date_ and _Last Used_. I believe that this is done to identify malicious devices with similar names.

Reply:
Thanks for the suggestion, added the `Added at` field.